### PR TITLE
Fix regex escaping

### LINKED ISSUE #86: Prompt Regex Escaping Confusion
First of all thanks for the new Prompt RegEx feature ❤️ 
I did just try it out and after some fiddeling around with the escaping, it works like a charm.

## Expected behaviour
First of all I think in some edge cases it  might make a difference to mention that the RegEx you type in your `config.py` is a JS RegEx and not a python one.
So that said IMHO it should work similar to this:

```python
copybutton_prompt_text = r"<valid_js_reg_ex> "
copybutton_prompt_is_regexp = True
```
🌈 🦄  

## What I did

In my case the prompt I wanted to get rid of was of the form 
```python
r"In \[\d*\]: |\.\.\.: "
```
which is the `jupyter console` style and pretty much the same as the iPython style from the docs 
```python 
"\\[\\d*\\]: |\\.\\.\\.: "
```

So it should be straight forward
```python 
"In \\[\\d*\\]: |\\.\\.\\.: "
```
as in the docs and I would be done?

But it didn't work, so I did some digging and ended up with:
```python 
"In \\[\\d*\\]: |\\.\\.\\.: |\\$ "
```

## What internally happens
In my `config.py`, I set `copybutton_prompt_text`
```python
copybutton_prompt_text = r"In \[\d*\]: |\.\.\.: "
```
which is the same as:
```python
copybutton_prompt_text = "In \\[\\d*\\]: |\\.\\.\\.: "
```
```python
>>> r"\\" == "\\\\"
True
```

Sphinx now replaces the context variable [`copybutton_prompt_text`](https://github.com/executablebooks/sphinx-copybutton/blob/d56f2cf69e7ad8346b9dad950b3cd03bdaba2c16/sphinx_copybutton/_static/copybutton.js_t#L89) in `copybutton.js_t` and generates `copybutton.js`.
But when we look at that string which should have just been passed as is, it is missing backslashes.
```javascript
'In \[\d*\]: |\.\.\.: |\$ '
```

Well this looks like a totaly valid JS RegEx, so this is fine?
And it would be if written like this (JS just wants keep people on their toes, so why not a different character for RegEx strings?):
```javascript
/In \[\d*\]: |\.\.\.: |\$ /
```

Now the string [gets passed to the RegExp constuctor](https://github.com/executablebooks/sphinx-copybutton/blob/d56f2cf69e7ad8346b9dad950b3cd03bdaba2c16/sphinx_copybutton/_static/copybutton_funcs.js#L14), which [expects strings to be escaped](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp#Flags_in_constructor) like python, if you don't use raw strings.
But it doesn't match.

```javascript
> "In [1]: bar".match(new RegExp('^(In \[\d*\]: |\.\.\.: )(.*)'))
null
```

#### TLDR
IMHO this all comes down ro JS RegExp 'helping' you in some cases when it can interprete what you might have ment, which gives you false security (false positiv match in the tests).

#### Details
Disclaimer I'm in no way a JS expert with insight into the internals, I just found this due to WTF moments + trail and error.

```javascript
> let proper_regex = new RegExp('^(>>> |\\$ |In \\[\\d*\\]: |\\[\\d*\\]: |\\.\\.\\.: )(.*)')
undefined
> let false_positive_regex = new RegExp('^(>>> |\\$ |In \[\d*\]: |\[\d*\]: |\.\.\.: )(.*)')
undefined
>"$ bar".match(proper_regex)
(3) ["$ bar", "$ ", "bar", index: 0, input: "$ bar", groups: undefined]
> "[1]: bar".match(proper_regex)
(3) ["[1]: bar", "[1]: ", "bar", index: 0, input: "[1]: bar", groups: undefined]
> "[1]: bar".match(false_positive_regex)
(3) ["[1]: bar", "[1]: ", "bar", index: 0, input: "[1]: bar", groups: undefined]
> "...: bar".match(proper_regex)
(3) ["...: bar", "...: ", "bar", index: 0, input: "...: bar", groups: undefined]
> "...: bar".match(false_positive_regex)
(3) ["...: bar", "...: ", "bar", index: 0, input: "...: bar", groups: undefined]
```
So far so good, all [testcases](https://github.com/executablebooks/sphinx-copybutton/blob/master/sphinx_copybutton/_static/test.js) work fine, as we would expect since they pass.

```javascript
> "In [1]: bar".match(proper_regex)
(3) ["In [1]: bar", "In [1]: ", "bar", index: 0, input: "In [1]: bar", groups: undefined]
> "In [1]: bar".match(false_positive_regex)
null
```

The only explaination I got for this behaviour, is that JS wants to "help" the user by guessing it's intention, like the famous adding thingy:
 ```javascript
> "5" + 1
"51"
```

but at some points it is like "I don't get it", and the pattern that worked before fails.

#### Use `repr` in `add_to_context`
If [this line](https://github.com/executablebooks/sphinx-copybutton/blob/d56f2cf69e7ad8346b9dad950b3cd03bdaba2c16/sphinx_copybutton/__init__.py#L19) was changed to:
  ```python
        {"copybutton_prompt_text": r"{}".format(repr(config.copybutton_prompt_text))[1: -1]}
  ```
I know it looks ugly, but would add extra backslashes for escaping (maybe someone else has a more elegant solution). 
I'm not sure how much [`escapeRegExp`](https://github.com/executablebooks/sphinx-copybutton/blob/d56f2cf69e7ad8346b9dad950b3cd03bdaba2c16/sphinx_copybutton/_static/copybutton_funcs.js#L1) would need to be adjusted to those changes.

I didn't have luck with backslash replacing in JS:
```javascript
> 'In \[\d*\]: |\.\.\.: '.replace(/\\/, '\\\\');
"In [d*]: |...: "
> 'In \[\d*\]: |\.\.\.: '
"In [d*]: |...: "
```

So in conclusion:
Python ❤️ 
JS 😠 

P.S.: Wish we had the `sphinx-copybutton` on github, for the JS blobs I posted 😢

### PR DESCRIPTION
**Pull request by s-weigand**
As concluded in #86, this fixes the way that `copybutton_prompt_text` is injected into `copybutton.js_t`.
And updates the docs on how to use the regex feature.
I left the old tests in there, since I kind like those oddities if you know them. But if that is too much clutter for you I can remove them.

Not sure how much you value [SemVer](https://semver.org/) but this is a breaking change, since it will break functionality for people that already use the regex feature. 

fixes #86 

**Note**: I left #84 ([besides](https://github.com/executablebooks/sphinx-copybutton/issues/84#issuecomment-645670875)) untouched, since IMHO this should configurable i.e. by a variable `copy_blank_lines`.  I just can see someone complaining around the corner, even so 'copy as is' is the most natural thing and should be default 😒

### CODE REVIEW

**Comment by choldgraf** [comment]
Thanks for this patch! @chrisjsewell do you think you could double-check that this looks correct? My regex skills are very poor 😬

File: doc/index.rst [hunk 1]
```diff
@@ -140,29 +140,66 @@ Using regexp prompt identifiers
...
-   copybutton_prompt_text = "\\[\\d*\\]: |\\.\\.\\.: "
+   copybutton_prompt_text = r">>> |\$ |\[\d*\]: |In \[\d*\]: |\.\.\.: "
    copybutton_prompt_is_regexp = True
 
-For example when using ipython prompts with continuations:
+Which also matches the prompts when using ipython prompts with continuations:
 
 .. code-block:: restructuredtext
 
+   Old style:
```
Comment (chrisjsewell) [inline doc/index.rst:167]: Do we actually need this old and new style? Was old style actually ever a thing (i.e. was it used in old versions of ipython) or can we just remove it (the examples and in the regex) and keep only the "new" style

File: sphinx_copybutton/_static/test.js [hunk 2]
```diff
@@ -105,20 +105,39 @@ $ second`,
...
+[2]: second
+In [3]: jupyter`,
+		prompt: '\\[\\d*\\]: |In \\[\\d*\\]: |\\.\\.\\.: ',
+		isRegexp: true,
+		onlyCopyPromptLines: true,
+		removePrompts: false,
+		expected: '[1]: first\n...: continuation\n[2]: second\nIn [3]: jupyter',
+	},
+	{
+	/* The following is included to demonstrate an example of a false positive regex test.
```
Comment (chrisjsewell) [inline sphinx_copybutton/_static/test.js:136]: Note, I changed this directly on your branch, so make sure to update your local checkout 😬

**Comment by chrisjsewell** [comment]
> Not sure how much you value SemVer but this is a breaking change, since it will break functionality for people that already use the regex feature.

Yep we should release this as 0.3.0

**Comment by s-weigand** [comment]
> 
> 
> > Not sure how much you value SemVer but this is a breaking change, since it will break functionality for people that already use the regex feature.
> 
> Yep we should release this as 0.3.0

Well, as for SemVer that would be a `1. 0. 0` as it might break things.IMHO SemVer only makes sense if it is consistent, which manually is hard to do and I didn't find the proper tool for it in the python universe / GH actions to automate it for me (maybe you can give me a pointer 😉).
I.e. in a project we had a dependency which changed from `0.9.9` to `1.0.0` and all was fine. 
But when they went to `1.0.1` many tests went down, so IMHO inconsistent versioning is more harmfull.
I think the easiest versioning scheme to maintain, is the date one as pip uses if for about 1.5y.

File: sphinx_copybutton/_static/test.js (hunk 2 above)
Comment (s-weigand) [reply sphinx_copybutton/_static/test.js:136]: Thx for reminding me 😄

File: doc/index.rst (hunk 1 above)
Comment (s-weigand) [reply doc/index.rst:167]: I also wasn't sure if the "old style" was ever like this, `ipython console`, `jupyter console` and `qt-console` use the "new style".
But I thought "there had to be a reason why `[1]` was used" (that's why I came up with the "old" and "new"). So If you can't think of an "old" use case, I will remove it.

**Comment by chrisjsewell** [comment]
> Well, as for SemVer that would be a 1. 0. 0

Yeh it's a bit of a cheat, but for versioning < 1.0.0, I generally treat MINOR version changes as also being backwards-incompatible, and PATCH for anything else. I think that's reasonable, since below 1.0 you're generally in "beta" mode where users should not yet expect the code to be completely "stable"

**Comment by choldgraf** [comment]
Yep I'm in the same boat - IMO pre `1.0` you treat "minor" versions as "major" versions, and pre `0.1` you treat "patch" versions as "major" versions

**Comment by s-weigand** [comment]
> 
> 
> > Well, as for SemVer that would be a 1. 0. 0
> 
> Yeh it's a bit of a cheat, but for versioning < 1.0.0, I generally treat MINOR version changes as also being backwards-incompatible, and PATCH for anything else. I think that's reasonable, since below 1.0 you're generally in "beta" mode where users should not yet expect the code to be completely "stable" anyway

Yeah, fair enough, versions below `1.0` are a gray area ^^

**Comment by chrisjsewell** [comment]
> maybe you can give me a pointer

Well I don't know it there can be a completely programmatic identifier of back-compatability breaking.
But I'm trying to push this commit convention now https://executablebooks.org/en/latest/dev-conventions.html#commit-messages, which hopefully will help to make it clearer where breaking changes occur,
i.e. the commit from this PR will start: `‼️ BREAKING: `

**Comment by s-weigand** [comment]
> 
> 
> > maybe you can give me a pointer
> 
> Well I don't know it there can be a completely programmatic identifier of back-compatability breaking.
> But I'm trying to push this commit convention now https://executablebooks.org/en/latest/dev-conventions.html#commit-messages, which hopefully will help to make it clearer where breaking changes occur,
> i.e. the commit from this PR will start: `‼️ BREAKING: `

In my TypeScript projects with template [typescript-library-starter](https://github.com/alexjoverm/typescript-library-starter) I use [git-cz](https://www.npmjs.com/package/git-cz), which basically wrappes git commits in a CLI dialog and adds keywords which get picked up further down the line. But 'completely programmatic identifier of back-compatability breaking' might be a dream.. enough off topic xD

**Comment by chrisjsewell** [comment]
> In my TypeScript projects with template typescript-library-starter I use git-cz

Yeh interesting cheers, there's a few different categories, but quite similar. I guess there's no convention that can cover all eventualities, but its better to have *a* convention than none at all.

> which basically wrappes git commits in a CLI dialog 

The only problem with any locally installed "commit tool" is that actually you're mainly writing the commit message directly here in the GitHub PR interface

File: doc/index.rst (hunk 1 above)
Comment (s-weigand) [reply doc/index.rst:167]: Also, whlie derping around with  `ipython`, `jupyter console` and `jupyter qtconsole`  I found differences with line continuation:

`ipython` and `jupyter qtconsole`:
```
In [1]: def test():
   ...:     return 1
```
`jupyter console`:
```
In [1]: def test():
       :     return 1
```
So the regexes should be `r"\s+\.\.\.: "` and `r"\s+: "`

Comment (s-weigand) [reply doc/index.rst:167]: What do you think about replacing: 
- "Old style" with "`ipython` and `qtconsole` style"
- "New style" with "`jupyter console` style"

Also using the regexes `r" {2,5}\.\.\.: "` and `r" {5,8}: "` would support continuations for `In []: ` up to `In [999]`, an be more robust against false positive matches than my proposed version with `\s+`.

Comment (chrisjsewell) [reply doc/index.rst:167]: didn't even know there was a jupyter console lol, but yeh sounds good to me if you want to have the two examples, or just use ipython

**Comment by chrisjsewell** [comment]
thanks @s-weigand!

**Comment by choldgraf** [comment]
thanks for the patch @s-weigand !!
